text/0034-ternary-hash.md

```diff
@@ -0,0 +1,472 @@
+This RFC relies on the presence of the types `TritBuf` and `Trits`, as defined by
+[RFC36](https://github.com/iotaledger/bee-rfcs/blob/master/text/0036-ternary.md), which are assumed to be owning and
+borrowing collections of binary-encoded ternary in the `T1B1` encoding (one trit per byte).
+
+```rust
+/// The common interface of cryptographic hash functions that follow the sponge construction and that act on ternary.
+trait Sponge {
+    /// An error indicating that a failure has occured during a sponge operation.
+    type Error;
+
+    /// Absorb `input` into the sponge.
+    fn absorb(&mut self, input: &Trits) -> Result<(), Self::Error>;
+
+    /// Reset the inner state of the sponge.
+    fn reset(&mut self);
+
+    /// Squeeze the sponge into a buffer
+    fn squeeze_into(&mut self, buf: &mut Trits) -> Result<(), Self::Error>;
+
+    /// Convenience function using `Sponge::squeeze_into` to return an owned version of the hash.
+    fn squeeze(&mut self) -> Result<TritBuf, Self::Error> {
+        let mut output = TritBuf::zeros(HASH_LENGTH);
+        self.squeeze_into(&mut output)?;
+        Ok(output)
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge into a buffer, and reset the sponge in one go.
+    fn digest_into(&mut self, input: &Trits, buf: &mut Trits) -> Result<(), Self::Error> {
+        self.absorb(input)?;
+        self.squeeze_into(buf)?;
+        self.reset();
+        Ok(())
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge, and reset the sponge in one go.
+    /// Returns an owned version of the hash.
+    fn digest(&mut self, input: &Trits) -> Result<TritBuf, Self::Error> {
+        self.absorb(input)?;
+        let output = self.squeeze()?;
+        self.reset();
+        Ok(output)
+    }
+}
+```
+
+Following the sponge metaphor, an input provided by the user is `absorb`ed, and an output will be `squeeze`d from the
+data structure. `digest` is a convenience method calling `absorb` and `squeeze` in one go. The `*_into` versions of
+these methods are for passing a buffer into which the calculated hashes are written. The internal state will not be
+cleared unless `reset` is called.
+
+### Design of `CurlP`
+
+`CurlP` is designed as a hash function that acts on a `T1B1` binary-encoded ternary buffer, with a hash length of `243`
+trits and an inner state of `729`:
+
+```rust
+const STATE_LENGTH: usize = HASH_LENGTH * 3;
+```
+
+In addition, a lookup table is used as part of the absorption step:
+
+||-1|0|1|
+|-|-|-|-|
+|**-1**|1|1|-1|
+|**0**|0|-1|1|
+|**1**|-1|0|0|
+
+Given two balanced trits `t` and `t'`, the following table can easily be accessed by shifting them to unbalanced trits:
+```rust
+const TRUTH_TABLE: [[i8; 3]; 3] = [[1, 0, -1], [1, -1, 0], [-1, 1, 0]];
+```
+
+The way `CurlP` is defined, it can not actually fail, because the input or outputs can be of arbitrary size; hence,
+the associated type `Error = Infallible`.
```

Comment:
I missed the line where the RFC clearly states, that a block `impl Sponge for CurlP` exists. It does provide it for the blanket impl, so I think it would be part of the "detailed design" to at least mention those impl blocks for CurlP and Kerl.

Reply:
Sorry I am not sure to understand this comment. And fail to see how it relates to the code you selected.

Reply:
Hmm I now understand given your next comment but tbh I don't really know what to think about. It's obvious impl will exist. I'm not sure the value to add them in the RFC since I'd definitely not impl the methods so it's basically writing the trait two more times.

Reply:
yea, I knew that response would coming (for sure it's trivial that the `Sponge` trait would be implemented on them), and I think intelligent readers will draw this conclusion immediatedly. It's your choice. Just wanted to mention it, that it *might* confuse some readers.

Reply:
Added